internal/destination/destination.go

```diff
@@ -131,42 +137,93 @@ func (d *Destination) Open(ctx context.Context) error {
 	if err != nil {
 		return fmt.Errorf("connect to NATS: %w", err)
 	}
-
-	d.writer, err = NewWriter(WriterParams{
-		Conn:          conn,
-		Subject:       d.config.Subject,
-		RetryWait:     d.config.RetryWait,
-		RetryAttempts: d.config.RetryAttempts,
-	})
-	if err != nil {
-		return fmt.Errorf("init jetstream writer: %w", err)
-	}
+	d.nc = conn
 
 	// Async handlers & callbacks
 	conn.SetErrorHandler(internal.ErrorHandlerCallback(ctx))
-	conn.SetDisconnectErrHandler(internal.DisconnectErrCallback(ctx, func(c *nats.Conn) {}))
-	conn.SetReconnectHandler(internal.ReconnectCallback(ctx, func(c *nats.Conn) {}))
+	conn.SetDisconnectErrHandler(internal.DisconnectErrCallback(ctx, func(c *nats.Conn) {
+		d.writer.stopWrites()
+	}))
+	conn.SetReconnectHandler(internal.ReconnectCallback(ctx, func(c *nats.Conn) {
+		d.writer, err = NewWriter(writerParams{
+			nc:            d.nc,
+			subject:       d.config.Subject,
+			retryWait:     d.config.RetryWait,
+			retryAttempts: d.config.RetryAttempts,
+		})
+		d.writer.startWrites()
+	}))
 	conn.SetClosedHandler(internal.ClosedCallback(ctx))
 	conn.SetDiscoveredServersHandler(internal.DiscoveredServersCallback(ctx))
 
+	d.writer, err = NewWriter(writerParams{
+		nc:            d.nc,
+		subject:       d.config.Subject,
+		retryWait:     d.config.RetryWait,
+		retryAttempts: d.config.RetryAttempts,
+	})
+	if err != nil {
+		return fmt.Errorf("init jetstream writer: %w", err)
+	}
+
 	return nil
 }
 
 // Write writes a record into a Destination.
-func (d *Destination) Write(_ context.Context, records []sdk.Record) (int, error) {
-	for i, record := range records {
-		if err := d.writer.Write(record); err != nil {
-			return i, fmt.Errorf("write: %w", err)
+func (d *Destination) Write(ctx context.Context, records []sdk.Record) (int, error) {
+	timeout, cancel := context.WithTimeout(ctx, d.config.RetryWait*time.Duration(d.config.RetryAttempts))
+	defer cancel()
+
+	attempts := 0
+	recorded := 0
+	for _, record := range records {
+		attempts++
+
+		select {
+		case <-ctx.Done():
+			err := ctx.Err()
+			sdk.Logger(ctx).Debug().
+				Int("record total", len(records)).
+				Int("record recorded", recorded).
+				Err(err).
+				Msg("write stopped by context before having all records recorded")
+
+			return recorded, nil
```

Comment:
[Watch out](https://github.com/ConduitIO/conduit-connector-sdk/blob/main/destination.go#L57):
> Write must return a non-nil error if it returns n < len(r).

Besides, I don't think we need to listen to `ctx.Done` since the timeout context already wraps `ctx`. Unless we remove the timeout, then of course we need to have this case.

Reply:
We still need to return an error, otherwise the SDK is going to regard this as a bug and [generate one for us](https://github.com/ConduitIO/conduit-connector-sdk/pull/80/files#diff-b0e314d1e24c22a62ecb712e94913d77bed15abfd75e80a0b61189c0abe17453R381).